common/metadata-helper.cpp

```diff
@@ -0,0 +1,324 @@
+#include "metadata-helper.h"
+
+#ifdef WIN32
+
+#include <Windows.h>
+#include <memory>
+#include <vector>
+#include <algorithm>
+#include <iterator>
+
+#include <regex>
+#include <iostream>
+#include <sstream>
+#include <functional>
+#include <map>
+#include <memory>
+
+#include <librealsense2/rs.hpp>
+
+#define MAX_KEY_LENGTH 255
+#define MAX_VALUE_NAME 16383
+
+#endif
+
+namespace rs2
+{
+#ifdef WIN32
+    struct device_id
+    {
+        std::string pid, mi, guid, sn;
+    };
+
+    inline std::string to_lower(std::string x)
+    {
+        transform(x.begin(), x.end(), x.begin(), tolower);
+        return x;
+    }
+
+    bool operator==(const device_id& a, const device_id& b)
+    {
+        return (to_lower(a.pid) == to_lower(b.pid) &&
+            to_lower(a.mi) == to_lower(b.mi) &&
+            to_lower(a.guid) == to_lower(b.guid) &&
+            to_lower(a.sn) == to_lower(b.sn));
+    }
+
+    class windows_metadata_helper : public metadata_helper
+    {
+    public:
+        static bool parse_device_id(std::string id, device_id* res)
+        {
+            static const std::regex regex("[\\s\\S]*pid_([0-9a-fA-F]+)&mi_([0-9]+)#[0-9]&([0-9a-fA-F]+)&[\\s\\S]*\\{([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})\\}[\\s\\S]*");
```

Comment:
Isn't "[\s\S]*" the same as ".*"? And is not needed at the beginning/end of the regex if we use regex_search() instead of regex_match()?
If we're using a-fA-F, then no need to use to_lower(), and then can pass the string in with 'const &'?

Reply:
> Isn't "[\s\S]" the same as "."? 

Apparently no, I had problems matching special symbols with `.*`

> And is not needed at the beginning/end of the regex if we use regex_search() instead of regex_match()

Make sense, I'll give it a try

> If we're using a-fA-F, then no need to use to_lower()

Ultimately, `to_lower` is only used for comparisons. If we switch all comparisons to `strcasecmp` as you suggested it may be redundant

Reply:
> > Isn't "[\s\S]" the same as "."?

> Apparently no, I had problems matching special symbols with .*

According to [this](https://stackoverflow.com/questions/35248876/whats-the-difference-between-s-s-and-in-java-regular-expressions/35248940), ". doesn't catch line terminators (like new line) by default" but I don't think that would cover your difficulties. I wonder what kind of special characters...

But OK, keep it at [/s/S] but please comment so the nice wise-guy like me won't try to "optimize" :)